rfd/0096-securing-docker-images.md

```diff
@@ -0,0 +1,323 @@
+image. 
+
+An example Dockerfile, assuming the Teleport Debian package is supplied by the 
+CI system, might look like something like:
+
+```Docker
+FROM debian:11 as dumb-init
+RUN apt update && apt-get download dumb-init && dpkg-deb -R dumb-init*.deb /opt/dumb-init
+
+FROM debian:11 as teleport
+COPY teleport*.deb
+RUN dpkg-deb -R teleport*.deb /opt/teleport
+
+# NOTE: the CC image supplies libc, libgcc and a few basic runtime libraries for us
+FROM gcr.io/distroless/cc-debian11
+COPY --from=dumb-init /opt/dumb-init/bin/dumb-init /bin
+COPY --from=teleport /opt/teleport/bin/* /bin
+ENTRYPOINT ["/bin/dumb-init", "teleport", "start", "-c", "/etc/teleport/teleport.yaml"]
+```
+> **NOTE:** This unpack-and-copy installation method is only appropriate for
+> packages with no complex installation requirements, like post-inbstall hooks. 
+>
+> Also note that for the sake of clarity I'm only including one dependency package. 
+> In the real distribution there would be multiple packages required.
+
+### Alternative builders
+
+As part of researching this RFD, I examined a couple of alternative ways to construct 
+the Teleport image. 
+
+ * **bazel, [distroless](https://github.com/GoogleContainerTools/distroless) and [rules_docker](https://github.com/bazelbuild/rules_docker)**: 
+   Given that the underlying distroless images are built using `bazel`, it should be 
+   possible to construct a custom image for Teleport in the same way. After some 
+   experimentation, I found that
+
+   1. the Debian package installation technique used by `rules_docker` is essentially 
+      a tweaked version of the extract-and-copy approach used by the `Dockerfile` 
+      above, and
+
+   2. There is a major [chicken-or-egg problem](https://github.com/GoogleContainerTools/distroless/issues/542)
+      in the build process when `distroless` is used as an external dependency in 
+      an enclosing `bazel` workspace, requiring manual intervention in the build to
+      solve.
+
+   Using `bazel` does not resolve a major limitation of using a basic 
+   `Dockerfile` (i.e. the `xcopy` style install) and introduces more complexity, in 
+   terms of both build process and tooling, so was rejected in favour of the 
+   `Dockerfile` approach.
+
+ * **[apko](https://github.com/chainguard-dev/apko)**: Apko is a tool for quickly 
+   building minimalist, reproducible Alpine linux images, using a declarative 
+   format. While I found the tool verly neat, the images it generates are still closer 
+   to a "debug" Distroless image. 
+
+   Using `apko` would also require us to build an Alpine linux package for Teleport 
+   to integrate it nto the build.
+
+   I seriously considered recommending `apko`, as it has some neat features (e.g.
+   automatically producting a SBOM as part of the construction process), but in the 
+   end I rejected it because of the extra software included in the resulting images.
+
```

Comment:
We _could_ consider rolling our own base image (with a normal Dockerfile, without bazel). This would have some major downsides (primarily the effort required to both build the tooling to create the base image and maintain it), but it could also have some non-technical business advantages:
* If we can solve some (or all) of the issues with other tools above then providing a secure, simple image to the open source community could help expand our brand.
* We could potentially integrate security fixes for upstream packages faster, showing our commitment to security to potential customers.
* We could implement features like SBOM reports into our toolchain which would add some transparency for both Gravitational and any customers using our Teleport image.
* By implementing the upstream image ourselves we can insure that all current and future products can easily adhere to our security standards, without reinventing the wheel for each one. This is doubly so when taking multi-arch images into account, which introduces a lot of additional technical problems for each image to solve.

I don't know that this route conforms to our high level strategic goals, but I thought it would be worth considering.